CLUSTER.md

```diff
@@ -0,0 +1,451 @@
+# Redis Cluster v2 Proposal
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Design](#design)
+    - [Terminology](#terminology)
+    - [High Level Overview](#high-level-overview)
+    - [The Topology Director](#the-topology-director)
+      - [Interfaces](#interfaces)
+    - [The Failover Coordinator](#the-failover-coordinator)
+        - [Interfaces](#interfaces)
+    - [The Node](#the-node)
+        - [Interfaces](#interfaces)
+- [Flows](#flows)
+    - [Membership and role management](#membership-and-role-management)
+        - [Cluster Creation](#cluster-creation)
+        - [Join](#join)
+        - [Remove Node](#remove-node)
+        - [Migrate A Node between Failover Coordinators](#migrate-a-node-between-failover-coordinators)
+    - [Failover](#failover)
+        - [Failover Coordinator Tick](#failover-coordinator-tick)
+    - [Resharding](#resharding)
+- [Flotilla Implementation Details](#flotilla-implementation-details)
+    - [Roles](#roles)
+    - [Node and Failover Coordinator IDs](#node-and-failover-coordinator-ids)
+    - [Publish/Subscribe](#publish/subscribe) 
+    - [Reference Implementation](#reference-implementation)
+        - [Road to Production](#road-to-production) 
+        - [Potential Enhancements](#potential-enhancements)
+<!-- /toc -->
+
+## Summary
+
+A new approach to Redis Clustering is proposed in order to:
+
+- Provide predictable behavior & correctness guarantees across the various possible failure modes
+- Create a minimal, well-defined internal interface to the clustering code so alternative approaches can easily replace the
+  entire clustering functionality at compile time.
+- Support a significantly higher number of shards
+- Significantly reduce failover time to under 10 seconds (at worse) and under 5 seconds in some configurations 
+- Allow for alternative consistency mode of operation
+- Long term: Support a clustered, but un-sharded deployment mode with no cross-slot limitations, which will allow us to deprecate all Sentinel code
```

Comment:
> > If by multi-tenancy you mean multi-database (with the SELECT command), that's only supported with unclustered Redis, and this new spec does not talk about adding it to clustered mode at the moment.
> 
> We do intend for this system to support un-sharded configurations though, in which we could / would support multiple databases and SELECT (and terminate Sentinel some day). @ushachar please ack, and see if the text needs any addition to make it clear.

Definitely ACK :)

Reply:
This doesn't have anything to do with pubsub -- it is needed to allow us to deprecate and eventually remove all the Sentinel code.
How it will be implemented (and whether it would be possible to transition between such unsharded to sharded modes) is completely TBD.

Reply:
Here, you mention that a clustered, but un-sharded deployment mode with no cross-slot limitations, can I understand that in a cluster, there are some nodes, but the data in these nodes is not sharded and there is no cross-slot data.  

That means in this cluster, there are 3 differents kinds of nodes： master node (only one node), some replica nodes, and some empty node.  And then when it is necessary, the replica nodes or empty node could transfer to as a sentinel mode node or you mean make some replica nodes or empty has sentinel function?   

I know this is a long-ternm plan, but do you have some draft idea or you already have it?  Thanks  @ushachar

Reply:
I see this is generating a lot of confusion, so will change the phrasing in the next version.
The idea is to allow replacing a Sentinel based deployment with a Flotilla based deployment.

Haven't spent a lot of time considering exactly how, but I'd imagine we would have the node(s) with TD+FC act as the Sentinel server(s) and have the primary + all replicas assigned to them.

Reply:
(hopefully) made the goal of this item a bit clearer